File: patches/experimental/udp-bytes/udp_connect_peer.js
function udp_connect_peer(socket,family,a,b,c,d,port,scope) {
  if((family!==4 && family!==6) || port===0 || port>65535 ||
      (family===4 && (b!==0 || c!==0 || d!==0 || scope!==0)))
    return io_tup(socket,io_fail(22));
  const sys=io_sys();
  const at=new Uint8Array(family===4?16:28);
  const view=new DataView(at.buffer);
  const little=new Uint8Array(new Uint16Array([1]).buffer)[0]===1;
  const af=family===4?2:(sys.mac?30:10);
  if(sys.mac){at[0]=at.length;at[1]=af;}else view.setUint16(0,af,little);
  view.setUint16(2,port,false);
  if(family===4)view.setUint32(4,a,false);
  else {
    [a,b,c,d].forEach((word,i)=>view.setUint32(8+4*i,word,false));
    view.setUint32(24,scope,little);
  }
  const result=sys.connect(socket,sys.ptr(at),at.length);
  return io_tup(socket,result<0?io_fail(sys.errno()):io_done({$:"Unit"}));
}

File: patches/experimental/udp-bytes/udp_connect_peer.c
// Associate an existing datagram socket with a numeric peer. UDP has no
// connection handshake; getsockname can then report the selected local route.
Term udp_connect_peer_run(Env e, Term* f, IoWork* w) {
  int fd=(int)io_hand_v(f[0]);
  u32 family=f[1],port=f[6],scope=f[7];
  if((family!=4 && family!=6) || port==0 || port>65535 ||
     (family==4 && (f[3] || f[4] || f[5] || scope)))
    return io_tup(e,io_hand(fd),io_fail(e,EINVAL,NULL));
  struct sockaddr_storage storage={0};
  socklen_t length;
  if(family==4) {
    struct sockaddr_in* at=(struct sockaddr_in*)&storage;
    at->sin_family=AF_INET; at->sin_port=htons(port);
    at->sin_addr.s_addr=htonl(f[2]); length=sizeof(*at);
  } else {
    struct sockaddr_in6* at=(struct sockaddr_in6*)&storage;
    at->sin6_family=AF_INET6; at->sin6_port=htons(port); at->sin6_scope_id=scope;
    uint32_t words[4]={htonl(f[2]),htonl(f[3]),htonl(f[4]),htonl(f[5])};
    memcpy(&at->sin6_addr,words,sizeof(words)); length=sizeof(*at);
  }
  int result=connect(fd,(struct sockaddr*)&storage,length);
  return io_tup(e,io_hand(fd),result<0 ? io_fail(e,errno,NULL) : io_done(e,term_pak(CID_UNIT,0)));
}
static void __attribute__((constructor)) udp_connect_peer_use(void) {
  io_eff(CID_UDP_CONNECT_PEER,udp_connect_peer_run,0);
}
